src/ruby/puppetserver-lib/puppet/server/catalog.rb

```diff
@@ -1,121 +1,124 @@
-                      request_data['trusted_facts']['values']
-                    else
-                      fact_values = facts.to_data_hash['values']
-                      fact_values['trusted']
-                    end
-    # If no trusted facts could be found, ensure a hash is returned
-    trusted_facts ||= {}
-  end
+        facts
+      end
 
-  def self.get_facts_from_pdb(nodename, environment)
-    if Puppet[:storeconfigs] == true
-      pdb_terminus = Puppet::Node::Facts::Puppetdb.new
-      request = Puppet::Indirector::Request.new(pdb_terminus.class.name,
-                                                :find,
-                                                nodename,
-                                                nil,
-                                                :environment => environment)
-      facts = pdb_terminus.find(request)
-
-      # If no facts have been stored for the node, PDB will return nil
-      if facts.nil?
-        # Create an empty facts object
-        facts = Puppet::Node::Facts.new(request_data['certname'])
+      def self.extract_trusted_facts(request_data, facts)
+        # Pull the trusted facts from the request, or attempt to extract them from
+        # the facts hash
+        trusted_facts = if request_data['trusted_facts']
+                          request_data['trusted_facts']['values']
+                        else
+                          fact_values = facts.to_data_hash['values']
+                          fact_values['trusted']
+                        end
+        # If no trusted facts could be found, ensure a hash is returned
+        trusted_facts ||= {}
       end
 
-      facts
-    else
-      # How should this be surfaced? Seems like we could maybe do better than a 500, unless
-      # that's accurate?
-      raise(Puppet::Error, "PuppetDB not configured, please provide facts with your catalog request.")
+      def self.get_facts_from_pdb(nodename, environment)
+        if Puppet[:storeconfigs] == true
+          pdb_terminus = Puppet::Node::Facts::Puppetdb.new
+          request = Puppet::Indirector::Request.new(pdb_terminus.class.name,
+                                                    :find,
+                                                    nodename,
+                                                    nil,
+                                                    :environment => environment)
+          facts = pdb_terminus.find(request)
+
+          # If no facts have been stored for the node, PDB will return nil
+          if facts.nil?
+            # Create an empty facts object
+            facts = Puppet::Node::Facts.new(request_data['certname'])
+          end
+
+          facts
+        else
+          # How should this be surfaced? Seems like we could maybe do better than a 500, unless
+          # that's accurate?
+          raise(Puppet::Error, "PuppetDB not configured, please provide facts with your catalog request.")
```

Comment:
Playing with this more, I think this is part of determining whether to call PDB in the first place and might work better in the caller?